### PR TITLE
Extract Jenkins into shared library

### PR DESCRIPTION
**Pull request by johnoliver**
Take some of the classes used in jenkins and extract them to a shared lib so that they can be reused.

Rename dir inside archive to be semver based on the parsed version.

Apply a bunch of the fixes also in #968

### CODE REVIEW

File: pipelines/build/common/openjdk_build_pipeline.groovy [hunk 1]
```diff
@@ -1,5 +1,5 @@
-import NodeHelper
-@Library('openjdk-jenkins-helper@master')
+@Library('local-lib@master')
+import common.*
```
Comment (karianna) [inline pipelines/build/common/openjdk_build_pipeline.groovy]: prefer explicit imports

File: pipelines/build/common/openjdk_build_pipeline.groovy [hunk 2]
```diff
@@ -375,7 +242,7 @@ class Build {
                         context.string(name: 'PRODUCT_MAJOR_VERSION', value: "${versionData.major}"),
                         context.string(name: 'PRODUCT_MINOR_VERSION', value: "${versionData.minor}"),
                         context.string(name: 'PRODUCT_MAINTENANCE_VERSION', value: "${versionData.security}"),
-                        context.string(name: 'PRODUCT_PATCH_VERSION', value: "${versionData.build}"),
+                        context.string(name: 'PRODUCT_PATCH_VERSION', value: "${String.format("%02d", versionData.build)}"),
```
Comment (karianna) [inline pipelines/build/common/openjdk_build_pipeline.groovy]: Could we ever have 3 decimals?

File: pipelines/library/src/ParseVersion.groovy [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+class ParseVersion {
+
+    static void main(String[] args) {
+        OptionAccessor parsedArgs = parseArgs(args)
+        VersionInfo version
+        if (parsedArgs.s) {
+            version = parseFromStdIn(args)
+        } else {
+            version = new VersionInfo().parse(args[args.length - 2], args[args.length - 1])
```
Comment (karianna) [inline pipelines/library/src/ParseVersion.groovy]: Could extract the 2 args into local vars that are named for readability.

File: pipelines/library/src/ParseVersion.groovy [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        }
+
+        printVersion(parsedArgs, version)
+    }
+
+    private static void printVersion(OptionAccessor parsedArgs, VersionInfo version) {
+        if (parsedArgs.f) {
```
Comment (karianna) [inline pipelines/library/src/ParseVersion.groovy]: `f`?

File: pipelines/library/src/ParseVersion.groovy [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        def reader = System.in.newReader()
+        def line
+        while ((line = reader.readLine()) != null) {
+            Matcher matcher = (line =~ /.*\(build (?<version>.*)\).*/)
+            if (matcher.matches()) {
+                VersionInfo version = new VersionInfo().parse(matcher.group("version"), args[args.length - 1])
+                return version
+            }
+        }
+        return null
```
Comment (karianna) [inline pipelines/library/src/ParseVersion.groovy]: What happens to the caller when they get a null?

File: pipelines/library/src/ParseVersion.groovy [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    private static OptionAccessor parseArgs(String[] args) {
+        CliBuilder cliBuilder = new CliBuilder()
+
+        cliBuilder.s('read input from stdin', args: 0)
```
Comment (karianna) [inline pipelines/library/src/ParseVersion.groovy]: `s` and `f`?

File: pipelines/library/src/common/VersionInfo.groovy [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    Integer minor // 0
+    Integer security // 181
+    Integer build // 8
+    String opt
+    String version
+    String pre
+    Integer adopt_build_number = 1
+    String semver
+
+    VersionInfo() {
```
Comment (karianna) [inline pipelines/library/src/common/VersionInfo.groovy]: Public constructor needed?

File: sbin/build.sh [hunk 8]
```diff
@@ -368,6 +371,17 @@ executeTemplatedFile() {
...
 }
 
+buildSharedLibs() {
+    cd "${LIB_DIR}"
+    ./gradlew clean uberjar
+}
+
+parseJavaVersionString() {
+  ADOPT_BUILD_NUMBER="${ADOPT_BUILD_NUMBER:-1}"
+  local version=$("$PRODUCT_HOME"/bin/java -version 2>&1 | java -cp "${LIB_DIR}/target/libs/pipelines-all-1.0.0-SNAPSHOT.jar" ParseVersion -s -f semver $ADOPT_BUILD_NUMBER)
```
Comment (karianna) [inline sbin/build.sh]: 1.0.0-SNAPSHOT seems brittle

File: sbin/build.sh [hunk 9]
```diff
@@ -408,47 +422,49 @@ printJavaVersionString()
...
+  mv "${jdkPath}" "${jdkTargetPath}"
 
   if [ -d "$(ls -d ${BUILD_CONFIG[JRE_PATH]})" ]
   then
-    JRE_TARGET_PATH="${openJdkVersion}-jre"
-    [ "${JRE_TARGET_PATH}" == "${openJdkVersion}" ] && JRE_TARGET_PATH="${openJdkVersion}.jre"
-    echo "moving $(ls -d ${BUILD_CONFIG[JRE_PATH]}) to ${JRE_TARGET_PATH}"
-    rm -rf "${JRE_TARGET_PATH}" || true
-    mv "$(ls -d ${BUILD_CONFIG[JRE_PATH]})" "${JRE_TARGET_PATH}"
+    [ "${jreTargetPath}" == "${openJdkVersion}" ] && jreTargetPath="${openJdkVersion}.jre"
```
Comment (karianna) [inline sbin/build.sh]: Should that be `jreTargetPath=="${openJdkVersion}.jre"` I'mconfused why an assignment occurs after an `&&`

File: pipelines/build/common/openjdk_build_pipeline.groovy (hunk 2 above)
Comment (johnoliver) [reply pipelines/build/common/openjdk_build_pipeline.groovy]: So this is a question re what the installer should expect to see here. For padding build numbers seem to get padded by 1 0, i.e `08`. 

However from what I understand of the installers they are using this info to build up what the directory inside the archive is. But under this new scheme the dir inside the archive is now `jdk-<semver>` so not padded at all. I doubt padded or not will fix this. This is a question for @gdams as to what should be provided here

File: pipelines/library/src/ParseVersion.groovy (hunk 4 above)
Comment (johnoliver) [reply pipelines/library/src/ParseVersion.groovy]: the sort of interesting way groovys CliBuilder works. If the user gave a `-f` argument then `parsedArgs.f` will exist

File: pipelines/library/src/ParseVersion.groovy (hunk 6 above)
Comment (johnoliver) [reply pipelines/library/src/ParseVersion.groovy]: same as before, this defines you can give arguments '-s' and '-f'

File: pipelines/library/src/common/VersionInfo.groovy (hunk 7 above)
Comment (johnoliver) [reply pipelines/library/src/common/VersionInfo.groovy]: groovy defaults to public

**Comment by karianna** [comment]
As an FYI to any watchers - we'll hold off landing this until after we release 12

File: pipelines/build/common/openjdk_build_pipeline.groovy (hunk 2 above)
Comment (gdams) [reply pipelines/build/common/openjdk_build_pipeline.groovy]: leave it without for now. If your fix is changing the directory name to be the semver version we should be good

**Comment by karianna** [comment]
@johnoliver Rebase.

**Pull request merged**